libraries/AP_LocationDB/AP_LocationDB.cpp

```diff
@@ -0,0 +1,414 @@
+    if (!healthy() || !is_valid_key(key)) {
+        return false;
+    }
+
+    {
+        WITH_SEMAPHORE(db_sem);
+        uint16_t item_idx;
+        // get index of item with given key from location database
+        if (get_item_index(key, item_idx)) {
+            // populate the item if the index is found and return true
+            ret = _items[item_idx];
+            return true;
+        }
+    }
+
+    return false;
+}
+
+// return true if we are able to retrieve item at a given index from the location database
+bool AP_LocationDB::get_item_at_index(const uint16_t index, AP_LocationDB_Item &ret) {
+    if (!healthy() || index >= _size) {
+        return false;
+    }
+
+    {
+        WITH_SEMAPHORE(db_sem);
+        // populate ret with the item at given index
+        ret = _items[index];
+    }
+
+    return true;
+}
+
+// return true if the item with given key is successfully removed from the location database
+bool AP_LocationDB::remove_item(const uint32_t key) {
+    if (!healthy() || !is_valid_key(key)) {
+        return false;
+    }
+
+    {
+        WITH_SEMAPHORE(db_sem);
+        uint16_t item_idx;
+        // get index of the item to be removed
+        if (get_item_index(key, item_idx)) {
+             _size -= 1; // decrease the size of database by one
+
+            if (_size != 0) {
+                // bring the last element at the place of item to be removed
+                // the items with index equal to or beyond database size are not considered to be a part of the database
+                _items[item_idx] = _items[_size];
+            }
+
+            return true;
+        }
+    }
+
+    return false;
+}
+
+// return true if we successfully update the item at given index of location database
+bool AP_LocationDB::update_item_at_index(const uint16_t index, AP_LocationDB_Item &new_item) {
+    // check of database is healthy and the key for the new item is valid
+    // also check if the passed index falls inside the current database size
+    if (!healthy() || !is_valid_key(new_item.get_key()) || index >= _size) {
+        return false;
+    }
+
+    {
+        WITH_SEMAPHORE(db_sem);
+        // copy new item at the place of the old item
+        _items[index] = new_item;
+    }
+
+    return true;
+}
+
+bool AP_LocationDB::update_item(const uint32_t key, AP_LocationDB_Item &new_item) {
+    if (!healthy() || !is_valid_key(key)) {
+        return false;
+    }
+
+#if !APM_BUILD_TYPE(APM_BUILD_UNKNOWN) && AP_AHRS_ENABLED
+    Vector3f item_pos;
+    Vector3f our_pos;
+    if (new_item.get_pos_cm_NEU(item_pos) && AP::ahrs().get_relative_position_NED_origin(our_pos)) {
+        our_pos.z = -our_pos.z;
+        item_pos = item_pos / 100; // cm to m
+        // check if the items is inside the inclusion radius
+        if ((item_pos - our_pos).length_squared() > (_inc_radius.get() * _inc_radius.get())) {
+            // item very far away from the vehicle
+            return false; 
```

Comment:
> Shouldn't we remove the item in this case?

If they are not updated for long enough they would automatically get timed out and hence removed. This gives them a small window to return back so that we do not have to add a new position for them if they just go out of the radius and return back quickly.



> ... also, shouldn't this check be in `update_item_at_index`?

We can move that but I think it would be good to check if the item is even inside the radius or not before iterating through the database and finding index for the element. That would just be a lot of extra work for no use if the item is outside the radius.